zio/ndjsonio/reader.go

```diff
@@ -6,28 +10,124 @@ import (
 )
 
 const (
 	ReadSize    = 64 * 1024
 	MaxLineSize = 50 * 1024 * 1024
 )
 
+type ReadStats struct {
+	*skim.Stats
+	*typeStats
+}
+
 type Reader struct {
 	scanner *skim.Scanner
-	parser  *Parser
+	inf     inferParser
+	typ     *typeParser
 	zctx    *resolver.Context
+	stats   ReadStats
 }
 
-func NewReader(reader io.Reader, zctx *resolver.Context) *Reader {
+func NewReader(reader io.Reader, zctx *resolver.Context) (*Reader, error) {
+	_, err := zctx.LookupTypeAlias("zenum", zng.TypeString)
+	if err != nil {
+		return nil, err
+	}
+
 	buffer := make([]byte, ReadSize)
+	scanner := skim.NewScanner(reader, buffer, MaxLineSize)
 	return &Reader{
-		scanner: skim.NewScanner(reader, buffer, MaxLineSize),
-		parser:  NewParser(zctx),
+		scanner: scanner,
+		stats:   ReadStats{Stats: &scanner.Stats, typeStats: &typeStats{}},
+		inf:     inferParser{zctx},
 		zctx:    zctx,
+	}, nil
+}
+
+// A Rule contains one or more matches and the name of a descriptor
+// key (in the companion Descriptors map).
+type Rule struct {
+	Match      map[string]string `json:"match"`
+	Descriptor string            `json:"descriptor"`
+}
+
+// A TypeConfig contains a map of Descriptors, keyed by name, and a
+// list of rules defining which records should be mapped into which
+// descriptor.
+type TypeConfig struct {
+	Descriptors   map[string][]interface{} `json:"descriptors"`
+	MatchingRules []Rule                   `json:"matching_rules"`
+}
+
+// typeRules is used internally and is derived from TypeConfig by
+// converting its descriptors into *zng.TypeRecord s for use by the
+// ndjson typed parser.
+type typeRules struct {
+	descriptors map[string]*zng.TypeRecord
+	rules       []Rule
+}
+
+// SetTypeConfig adds a TypeConfig to the reader. Its use is optional,
+// but if used, it should be called before records are processed.  In
+// the absence of a TypeConfig, records are all parsed with the
+// inferParser. If a TypeConfig is present, records are first parsed
+// with the typeParser, and if that fails, with the inferParser.
+func (r *Reader) SetTypeConfig(tc TypeConfig) error {
+	tr := typeRules{
+		descriptors: make(map[string]*zng.TypeRecord),
+		rules:       tc.MatchingRules,
+	}
+
+	for key, columns := range tc.Descriptors {
+		typeName, err := zjsonio.DecodeType(columns)
+		if err != nil {
+			return fmt.Errorf("error decoding type \"%s\": %s", typeName, err)
+		}
+		typ, err := r.zctx.LookupByName(typeName)
+		if err != nil {
+			return err
+		}
+		recType, ok := typ.(*zng.TypeRecord)
+		if !ok {
+			return fmt.Errorf("type not a record: \"%s\"", typeName)
+		}
+		tr.descriptors[key] = recType
+	}
+	r.typ = &typeParser{zctx: r.zctx, tr: tr, stats: r.stats.typeStats}
+	return nil
+}
+
+// Parse returns a zng.Encoding slice as well as an inferred zng.Type
+// from the provided JSON input. The function expects the input json to be an
+// object, otherwise an error is returned.
+func (r *Reader) Parse(b []byte) (zcode.Bytes, zng.Type, error) {
```

Comment:
nit: signature here would be a little simpler if this returned a zng.Value

Reply:
this is pre-existing but agreed; will change

Reply:
done